File: pkg/common/stats/connection.go
package stats

import (
	"sync"
	"sync/atomic"
	"time"

	"github.com/puppy/pkg/common"
)

// ConnectionInfo describes a single active tunnel observed by a frontend.
type ConnectionInfo struct {
	// ID is a unique identifier assigned when the connection is registered.
	ID string
	// Frontend is the name of the frontend that accepted the connection.
	Frontend string
	// RemoteAddr is the client address (host:port).
	RemoteAddr string
	// Target is the destination the backend dialed on behalf of the client.
	Target common.Target
	// Protocol is the detected application protocol (may be ProtocolUnknown).
	Protocol common.Protocol
	// Network is the transport network ("tcp" or "udp").
	Network string
	// StartedAt is when the connection was accepted.
	StartedAt time.Time
	// ClosedAt is set when the connection is removed from the registry.
	ClosedAt time.Time

	// bytesIn/bytesOut are mutated atomically via AddBytesIn/AddBytesOut by
	// the counting wrapper and are read atomically by BytesIn/BytesOut, so no
	// external locking is required even while the shim's two copy goroutines
	// run concurrently.
	bytesIn  atomic.Uint64
	bytesOut atomic.Uint64
}

// BytesIn returns the cumulative number of bytes read from the client.
func (c *ConnectionInfo) BytesIn() uint64 { return c.bytesIn.Load() }

// BytesOut returns the cumulative number of bytes written to the client.
func (c *ConnectionInfo) BytesOut() uint64 { return c.bytesOut.Load() }

// AddBytesIn atomically adds n to the connection's inbound byte counter. It
// is called by the counting connection wrapper; callers do not need external
// locking.
func (c *ConnectionInfo) AddBytesIn(n int) {
	if n <= 0 {
		return
	}
	c.bytesIn.Add(uint64(n))
}

// AddBytesOut atomically adds n to the connection's outbound byte counter.
func (c *ConnectionInfo) AddBytesOut(n int) {
	if n <= 0 {
		return
	}
	c.bytesOut.Add(uint64(n))
}

// ConnectionRegistry tracks active connections across all frontends. The set
// of active connections is expected to be small relative to total throughput,
// so a single RWMutex-guarded map is sufficient.
type ConnectionRegistry struct {
	mu     sync.RWMutex
	active map[string]*ConnectionInfo
}

// NewConnectionRegistry returns a ready-to-use registry.
func NewConnectionRegistry() *ConnectionRegistry {
	return &ConnectionRegistry{active: make(map[string]*ConnectionInfo)}
}

// Register records a new active connection and returns a pointer to the
// stored ConnectionInfo. The caller should use the returned pointer with a
// counting connection wrapper and call Remove when the connection closes.
func (r *ConnectionRegistry) Register(info *ConnectionInfo) *ConnectionInfo {
	if r == nil || info == nil {
		return nil
	}
	if info.StartedAt.IsZero() {
		info.StartedAt = time.Now()
	}
	r.mu.Lock()
	r.active[info.ID] = info
	r.mu.Unlock()
	return info
}

// Remove deletes a connection from the active set and marks ClosedAt.
func (r *ConnectionRegistry) Remove(id string) {
	if r == nil {
		return
	}
	r.mu.Lock()
	if info, ok := r.active[id]; ok {
		info.ClosedAt = time.Now()
		delete(r.active, id)
	}
	r.mu.Unlock()
}

// Get returns the connection with the given id, or nil if not found.
func (r *ConnectionRegistry) Get(id string) *ConnectionInfo {
	if r == nil {
		return nil
	}
	r.mu.RLock()
	info := r.active[id]
	r.mu.RUnlock()
	return info
}

// Active returns a snapshot slice of all currently active connections. The
// slice is a copy; callers may iterate it without holding the lock.
func (r *ConnectionRegistry) Active() []*ConnectionInfo {
	if r == nil {
		return nil
	}
	r.mu.RLock()
	out := make([]*ConnectionInfo, 0, len(r.active))
	for _, info := range r.active {
		out = append(out, info)
	}
	r.mu.RUnlock()
	return out
}

// ActiveByFrontend returns active connections for a specific frontend name.
func (r *ConnectionRegistry) ActiveByFrontend(frontend string) []*ConnectionInfo {
	if r == nil {
		return nil
	}
	r.mu.RLock()
	out := make([]*ConnectionInfo, 0)
	for _, info := range r.active {
		if info.Frontend == frontend {
			out = append(out, info)
		}
	}
	r.mu.RUnlock()
	return out
}

// Count returns the number of currently active connections.
func (r *ConnectionRegistry) Count() int {
	if r == nil {
		return 0
	}
	r.mu.RLock()
	n := len(r.active)
	r.mu.RUnlock()
	return n
}

File: rust/crates/puppy-core/src/stats/connection.rs
//! Per-connection info and the active-connection registry.

use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::time::Instant;

use parking_lot::RwLock;

use crate::backend::{Protocol, Target};

/// Describes a single active tunnel observed by a frontend.
///
/// Byte counters are atomic so the shim's two copy tasks can update them
/// concurrently without external locking.
#[derive(Debug)]
pub struct ConnectionInfo {
	/// Unique identifier assigned when the connection is registered.
	pub id: String,
	/// Name of the frontend that accepted the connection.
	pub frontend: String,
	/// Client address (`host:port`).
	pub remote_addr: String,
	/// Destination the backend dialed on behalf of the client.
	pub target: Target,
	/// Detected application protocol (may be `Unknown`).
	pub protocol: Protocol,
	/// Transport network (`"tcp"` or `"udp"`).
	pub network: String,
	/// When the connection was accepted.
	pub started_at: Instant,
	/// Set when the connection is removed from the registry. Shared via
	/// `RwLock<Option<Instant>>` so `Remove` can mutate it through an `Arc`.
	pub closed_at: RwLock<Option<Instant>>,

	bytes_in: AtomicU64,
	bytes_out: AtomicU64,
}

impl ConnectionInfo {
	/// Creates a new `ConnectionInfo` with zeroed byte counters and
	/// `started_at = Instant::now()`.
	pub fn new(
		id: impl Into<String>,
		frontend: impl Into<String>,
		remote_addr: impl Into<String>,
	) -> Self {
		Self {
			id: id.into(),
			frontend: frontend.into(),
			remote_addr: remote_addr.into(),
			target: Target {
				network: String::new(),
				protocol: Protocol::Unknown,
				host: String::new(),
				port: 0,
			},
			protocol: Protocol::Unknown,
			network: String::new(),
			started_at: Instant::now(),
			closed_at: RwLock::new(None),
			bytes_in: AtomicU64::new(0),
			bytes_out: AtomicU64::new(0),
		}
	}

	/// Creates a new `ConnectionInfo` with the given target/protocol/network
	/// and zeroed byte counters.
	pub fn with_target(
		id: impl Into<String>,
		frontend: impl Into<String>,
		remote_addr: impl Into<String>,
		target: Target,
		protocol: Protocol,
		network: impl Into<String>,
	) -> Self {
		Self {
			id: id.into(),
			frontend: frontend.into(),
			remote_addr: remote_addr.into(),
			protocol,
			network: network.into(),
			target,
			started_at: Instant::now(),
			closed_at: RwLock::new(None),
			bytes_in: AtomicU64::new(0),
			bytes_out: AtomicU64::new(0),
		}
	}

	/// Cumulative bytes read from the client.
	pub fn bytes_in(&self) -> u64 {
		self.bytes_in.load(Ordering::Relaxed)
	}

	/// Cumulative bytes written to the client.
	pub fn bytes_out(&self) -> u64 {
		self.bytes_out.load(Ordering::Relaxed)
	}

	/// Atomically adds `n` to the inbound byte counter. No-op for `n == 0`.
	pub fn add_bytes_in(&self, n: usize) {
		if n == 0 {
			return;
		}
		self.bytes_in.fetch_add(n as u64, Ordering::Relaxed);
	}

	/// Atomically adds `n` to the outbound byte counter. No-op for `n == 0`.
	pub fn add_bytes_out(&self, n: usize) {
		if n == 0 {
			return;
		}
		self.bytes_out.fetch_add(n as u64, Ordering::Relaxed);
	}

	/// Returns `true` if `closed_at` has been set.
	pub fn is_closed(&self) -> bool {
		self.closed_at.read().is_some()
	}
}

/// Tracks active connections across all frontends.
///
/// Connections are stored as `Arc<ConnectionInfo>` so external holders
/// (counting wrapper, tests) observe `Remove`'s `closed_at` mutation.
pub struct ConnectionRegistry {
	active: RwLock<std::collections::HashMap<String, Arc<ConnectionInfo>>>,
}

impl ConnectionRegistry {
	/// Returns a ready-to-use registry.
	pub fn new() -> Self {
		Self {
			active: RwLock::new(std::collections::HashMap::new()),
		}
	}

	/// Records a new active connection. The registry stores the `Arc` clone so
	/// external holders and the registry share the same `ConnectionInfo`.
	pub fn register(&self, info: Arc<ConnectionInfo>) -> Arc<ConnectionInfo> {
		let id = info.id.clone();
		self.active.write().insert(id, info.clone());
		info
	}

	/// Deletes a connection from the active set and marks `closed_at`.
	pub fn remove(&self, id: &str) {
		if let Some(info) = self.active.write().remove(id) {
			*info.closed_at.write() = Some(Instant::now());
		}
	}

	/// Returns the connection with the given id, or `None` if not found.
	pub fn get(&self, id: &str) -> Option<Arc<ConnectionInfo>> {
		self.active.read().get(id).cloned()
	}

	/// Returns a snapshot of all currently active connections.
	pub fn active(&self) -> Vec<Arc<ConnectionInfo>> {
		self.active.read().values().cloned().collect()
	}

	/// Returns active connections for a specific frontend name.
	pub fn active_by_frontend(&self, frontend: &str) -> Vec<Arc<ConnectionInfo>> {
		self.active
			.read()
			.values()
			.filter(|info| info.frontend == frontend)
			.cloned()
			.collect()
	}

	/// Returns the number of currently active connections.
	pub fn count(&self) -> usize {
		self.active.read().len()
	}
}

impl Default for ConnectionRegistry {
	fn default() -> Self {
		Self::new()
	}
}
